### PR TITLE
Diophantine update

### LINKED ISSUE #8943: diophantine misses trivial solution
```py
In [6]: diophantine((3*(x**2 + y**2 + z**2) - 14*(x*y + y*z + z*x)))
Out[6]: set([])
```

See http://math.stackexchange.com/questions/1133888/hard-number-theory-problem. The only solution is `(0, 0, 0)`. Nothing in the docstring indicates that it should skip trivial solutions.

### LINKED ISSUE #9538: diophantine() doesn't let you specify the variable order
`diophantine` gives a tuple of solutions, but it doesn't let you specify the variable order (it orders them alphabetically, I think). It should either require you to specify an order or return a dictionary. 

@thilinarmtb

### LINKED ISSUE #9539: diophantine(6*k + 9*n + 20*m - x) gives TypeError: unsupported operand type(s) for *: 'NoneType' and 'Symbol'
```
In [15]: diophantine(6*k + 9*n + 20*m - x)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-234fff11d03b> in <module>()
----> 1 diophantine(6*k + 9*n + 20*m - x)

/Users/aaronmeurer/Documents/python/sympy/sympy/sympy/solvers/diophantine.py in diophantine(eq, param)
     85
     86         var_t, jnk, eq_type = classify_diop(base)
---> 87         solution = diop_solve(base, param)
     88
     89         if eq_type in ["linear", "homogeneous_ternary_quadratic", "general_pythagorean"]:

/Users/aaronmeurer/Documents/python/sympy/sympy/sympy/solvers/diophantine.py in diop_solve(eq, param)
    173
    174     if eq_type == "linear":
--> 175         return _diop_linear(var, coeff, param)
    176
    177     elif eq_type == "binary_quadratic":

/Users/aaronmeurer/Documents/python/sympy/sympy/sympy/solvers/diophantine.py in _diop_linear(var, coeff, param)
    402         for v in var[2:]:
    403             sol_x, sol_y  = base_solution_linear(-coeff[v], a, b)
--> 404             X.append(sol_x*v)
    405             Y.append(sol_y*v)
```

If I reorder the variables it works.

```

In [14]: diophantine(6*n + 9*m + 20*k - x)
Out[14]: set([(24⋅n - 9⋅t - 4⋅x, -54⋅n + 20⋅t + 9⋅x, n, x)])

```

All variables are set as `integer=True`.

### LINKED ISSUE #10999: diop: holzer error
```
>>> holzer(2, 6, 2, 1, 1, 10)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/solvers/diophantine.py", line 2276, in holzer
    w = -(a*u_0*x_0 + b*v_0*y_0) // (c*z_0)
TypeError: unsupported operand type(s) for *: 'One' and 'NoneType'
```

addressed by #10997

### LINKED ISSUE #11000: diop: power_representation
```
>>> list(power_representation(1, 1, 1)) 
[()] 
```
This should be [(1,)]

addressed by #10997

### LINKED ISSUE #11016: diop: sum of squares needs to try more options to satisfy conditions
```python
>>> eq=Add(*[i**2 for i in symbols('x:6',nonnegative=1)])
>>> eq.subs(zip(eq.free_symbols,(10, 3, 0, 1, 1, 1)))
112
>>>
>>> diophantine(eq - 112)
set([])
```

Currently, only a single solution is returned to diophantine, (-10, -3, 0, 1, 1, 1, 1), but that one doesn't satisfy the condition that the values be nonnegative, so the empty set is returned rather than the solution without negatives.

### LINKED ISSUE #11021: diop: power_representation(4**5, 3, 1) -> (4,)
```python
>>> from sympy.solvers.diophantine import power_representation as f
>>> list(f(4**5,3,1))
[(4,)]
```

The above answer is not right since `4**3 != 4**5`.

Addressed by #10997

### LINKED ISSUE #11026: diophantine(x**3+y**3-2) -> KeyError instead of {(1, 1)}
```
>>> diophantine(x**3+y**3-2)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/solvers/diophantine.py", line 86, in diophantine
    var_t, jnk, eq_type = classify_diop(base)
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/solvers/diophantine.py", line 328, in classify_diop
    coeff[term] == Integer(0)
KeyError: x**2*y
```

### LINKED ISSUE #11049: diop: recursion error
```
>>> diophantine(92*x**2 - 99*y**2 - z**2)
  File "sympy\solvers\diophantine.py", line 671, in extended_euclid
    x0, y0, d = extended_euclid(b, a%b)
  File "sympy\solvers\diophantine.py", line 668, in extended_euclid
    if b == 0:
RuntimeError: maximum recursion depth exceeded in cmp
```

addressed by #10997

### LINKED ISSUE #11050: diop: partition(n, k) gives redundant result
```python
>>> from sympy.solvers.diophantine import partition
>>> list(partition(3))
[(1,1,1),(1,2),(3)] 
>>> list(partition(3,2))
[(2,1),(1,2)] 
```

That last result should just be [(2,1)]. This is fixed in #10997

### PR DESCRIPTION
**Pull request by smichr**
100% coverage with 1 pragma-no-cover.

fixes #8943
fixes #9538
closes #9539
fixes #10999 
fixes #11000 
fixes #11016 
fixes #11021 
fixes #11026
fixes #11049
fixes #11050

### CODE REVIEW

File: sympy/utilities/iterables.py [hunk 1]
```diff
@@ -1381,6 +1381,12 @@ def partitions(n, m=None, k=None, size=False):
         raise ValueError("n must be >= 0")
     if m == 0:
         raise ValueError("m must be > 0")
+    if n == 0:
+        if size:
+            yield 1, {0: 1}
```
Comment (jksuom) [inline sympy/utilities/iterables.py]: I'm not sure about these. Normally partitions are expected to consist of positive integers. Maybe empty dictionary could be returned?

Comment (smichr) [inline sympy/utilities/iterables.py]: I would feel better about not even allowing partitions of zero. What do you think about changing the error above to be "n must be > 0"?

Comment (jksuom) [inline sympy/utilities/iterables.py]: I think that is the best solution. There is no real reason to consider partitions of zero.

**Comment by asmeurer** [comment]
Don't know what your aims are here, but I have found a [couple of issues](https://github.com/sympy/sympy/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+author%3Aasmeurer+diophantine) I've found with diophantine in the past that don't appear to be addressed here.

**Comment by smichr** [comment]
> I have found a couple of issues

Those (and others) have now been addressed.

**Comment by smichr** [comment]
ping @thilinarmtb or anyone else that would give this a once over. Many issues are fixed. If there is no objection, I will commit in 96 hours.

**Comment by thilinarmtb** [comment]
@smichr : I'll go through this during the weekend.

**Comment by smichr** [comment]
@thilinarmtb , thanks. One thing I changed was the use of Integer vs int. Only Diophantine returns Integers; all other routines return an int. I know there was some talk about only exposing Diophantine. That seems like a good idea to me. I think I would also expose `power_representation` and/or the alias `sum_of_Powers`) and `sum_of_squares`.

I am also not sure about the need of keeping `sum_of_three_squares` and `sum_of_four_squares` except that if someone just wants a single instance, these are a quick way of getting it.

Finally, I tested loops over integers seeing when it might be more advantageous to use "smart" writing as squares instead of the backtracking recursive approach that is used by `power_representation`. I corrected the routine to watch for cases like writing 9 squares as a square and *two groups of four*, even wrote my own routines that did essentially the same thing, but it always seemed faster to just use the recursive approach. I'm a little doubtful that it truly is the best way to go; it seems like at some point the recursive approach would get slow and the "schorn" reference suggests the same, but everything I tested seemed to show that the recursive approach used here is best.

File: sympy/solvers/diophantine.py [hunk 2]
```diff
@@ -1827,65 +1839,52 @@ def _transformation_to_normal(var, coeff):
...
+        return Matrix(3, 3, [1, S(-B)/(2*A), S(-C)/(2*A), 0, 1, 0, 0, 0, 1])*T_0
 
-                else:
-                    # Ax**2 + E*y*z + F*z**2  = 0
-                    _var[0], _var[2] = var[2], var[0]
-                    T = _transformtion_to_normal(_var, coeff)
-                    T.row_swap(0, 2)
-                    T.col_swap(0, 2)
-                    return T
+    elif coeff[y*z] != 0:
```
Comment (thilinarmtb) [inline sympy/solvers/diophantine.py]: I guess we can simplify this as in other cases as well.

File: sympy/solvers/diophantine.py [hunk 3]
```diff
@@ -2062,58 +2062,61 @@ def _diop_ternary_quadratic_normal(var, coeff):
...
 
 
-def square_factor(a):
+def sqf_normal(a, b, c, steps=False):
     """
-    Returns an integer `c` s.t. `a = c^2k, \ c,k \in Z`. Here `k` is square
-    free.
+    Return `a', b', c'`, the coefficients of the square-free normal
+    form of `ax^2 + by^2 + cz^2 = 0`, where `a', b', c'` are pairwise
+    prime and `\gcd(a', b', c')` is `1`.  If `steps` is True then
```
Comment (thilinarmtb) [inline sympy/solvers/diophantine.py]: if `a, b, c` are pairwise prime, then `gcd(a, b, c) ` is 1. Do we need to restate it?

File: sympy/solvers/diophantine.py [hunk 4]
```diff
@@ -2262,8 +2261,6 @@ def descent(A, B):
         return (1, 0, 1)
     if A == 1:
         return (1, 1, 0)
-    if B == -1:
-        return (None, None, None)
```
Comment (thilinarmtb) [inline sympy/solvers/diophantine.py]: Why is this line removed?

File: sympy/solvers/diophantine.py [hunk 5]
```diff
@@ -2341,33 +2341,67 @@ def norm(u, w, a, b):
...
+                w += 1
+            assert abs(w - r) <= S.One
+
+        A = (a*u**2 + b*v**2 + c*w**2)
+        B = (a*u*x_0 + b*v*y_0 + c*w*z_0)
+        x = (x_0*A - 2*u*B) / k
+        y = (y_0*A - 2*v*B) / k
+        z = (z_0*A - 2*w*B) / k
+        assert all(i.is_Integer for i in (x, y, z))
+        step += 1
```
Comment (thilinarmtb) [inline sympy/solvers/diophantine.py]: Looks like this is a different implementation  than the existing one.

File: doc/src/modules/solvers/diophantine.rst [hunk 6]
```diff
@@ -281,7 +281,7 @@ level API.
 >>> diophantine(a**2 + b**2 + c**2 + d**2 + e**2 + f**2 - 112)
 set([(-10, -3, 0, 1, 1, 1)])
 
-If you want to get a more thorough idea about the the Diophantine module please
+If you want to get a more thorough understanding of the Diophantine module please
 refer to the following blog.
```
Comment (thilinarmtb) [inline doc/src/modules/solvers/diophantine.rst]: I guess `refer the following blog` sounds better.

**Comment by thilinarmtb** [comment]
@smichr : I went through the PR and you have done an amazing job of polishing the code and documentation. Thanks for your work ! I left a few comments.

> I know there was some talk about only exposing Diophantine. That seems like a good idea to me. I think I would also expose power_representation and/or the alias sum_of_powers) and sum_of_squares.

Yes, if we expose other routines, there can be a lot of confusion. Why would you expose the sum of powers?

> I am also not sure about the need of keeping sum_of_three_squares and sum_of_four_squares except that if someone just wants a single instance, these are a quick way of getting it.

IMO, we shouldn't expose these two. `diop_general_sum_of_squares` check for these two cases upfront so calling them directly won't have any significant impact.

> Finally, I tested loops over integers seeing when it might be more advantageous to use "smart" writing as squares instead of the backtracking recursive approach that is used by power_representation. I corrected the routine to watch for cases like writing 9 squares as a square and two groups of four, even wrote my own routines that did essentially the same thing, but it always seemed faster to just use the recursive approach. I'm a little doubtful that it truly is the best way to go; it seems like at some point the recursive approach would get slow and the "schorn" reference suggests the same, but everything I tested seemed to show that the recursive approach used here is best.

I see. When I implemented it, even I expected it to be slow for larger values.

**Comment by smichr** [comment]
I found a reference for counting the ways to write `n` as a sum of 2, 3 or 4 squares [here](http://web.maths.unsw.edu.au/~mikeh/webpapers/paper73.pdf) but I don't know what the sigma function is in equation (6). Can you help me? I also found a reference that refers to the [Jacobi formula](http://kam.mff.cuni.cz/~klazar/ostrava1.pdf). Is sigma(n) the sum of the divisors of n?

**Comment by thilinarmtb** [comment]
Yes, I think it is the sum of divisors of n.

Take the last formula in the page 3 of your first [reference](http://web.maths.unsw.edu.au/~mikeh/webpapers/paper73.pdf) and the equation 1 in the second [reference] (http://kam.mff.cuni.cz/~klazar/ostrava1.pdf) (Jacobi formula paper). They are the same as LHSs are the same in both the equations (Only the notation is different, `\phi(q)` instead of the summation).

Lemma 4 [here](http://www.rowan.edu/colleges/csm/departments/math/facultystaff/osler/89%20Surprising%20Connection%20Between%20Partitions%20and%20Divisors.pdf.pdf), indicates that `\sigma(n)` in the first reference is indeed the sum of divisors. Equation 1 in the second reference is just a rewrite of the equation in the first reference using the Lemma 4.

File: sympy/solvers/diophantine.py (hunk 2 above)
Comment (smichr) [inline sympy/solvers/diophantine.py]: I'm sorry, I'm not sure in which way you mean this might be simplified.

File: sympy/solvers/diophantine.py (hunk 3 above)
Comment (smichr) [inline sympy/solvers/diophantine.py]: I removed that phrase now.

File: sympy/solvers/diophantine.py (hunk 4 above)
Comment (smichr) [inline sympy/solvers/diophantine.py]: It is already handled by the routine. But the case of `A = B = -1` leads to infinite recursion without that; I added it back with a test.

File: doc/src/modules/solvers/diophantine.rst (hunk 6 above)
Comment (smichr) [inline doc/src/modules/solvers/diophantine.rst]: It sounds fine as it is (to me). Do you want me to shorten it further to "More discussion about the Diophantine module can be found at the following blog:"?

File: sympy/solvers/diophantine.py (hunk 5 above)
Comment (smichr) [inline sympy/solvers/diophantine.py]: Yes, there were some issues with the former implementation. I just reinterpreted the reference and what I've committed is that interpretation.  Here are some examples of failures in the former version:

```
>>> f = lambda eq: _diop_ternary_quadratic_normal((x, y, z), eq.as_coefficients_dict())
>>> f(92*x**2 - 99*y**2 - z**2)
(28557, 21296, 155058)  <-- should be set([(9, 7, 51)])
>>> f(234*x**2 - 65601*y**2 - z**2)
(5945, 31, 126819) <-- should be set([(233, 1, 3555)])
```

Comment (thilinarmtb) [inline sympy/solvers/diophantine.py]: Great !!!

**Comment by smichr** [comment]
> diop_general_sum_of_squares check for these two cases upfront

This no longer uses those routines since they only returned one particular instance; now all solutions are returned.

Regarding `partition`: you found a great reference for a routine but that aspect of the original `partition` function was not used (the case when all partitions should be generated) and the issue I opened indicated that there were problems. As a quick fix, I just replaced it with something that would expand the dictionaries received from `partitions`; I have since found a nice way to use that reference, however, and will update this PR with it. Alternatively, `sum_of_powers` could be made to return a dictionary, too, and then it could just use the `partitions` function directly.

**Comment by thilinarmtb** [comment]
That's great !! Which reference are you talking about? Did I implement that wrong?

**Comment by smichr** [comment]
> Which reference are you talking about? Did I implement that wrong?

The reference is [this one](http://jeromekelleher.net/generating-integer-partitions.html); there must be some problem (see #11050).

**Comment by smichr** [comment]
> I expected it to be slow for larger values

I think part of the reason it is faster than expected is because you are stepping like  int(sqrt(10)), int(sqrt(10//2)), int(sqrt(10//3)), ... rather than blindly starting from int(sqrt(10)) and going down by 1 each time. As that n=10 gets to be bigger, this has a greater impact.

**Comment by smichr** [comment]
I'm just recording here the routines for counting the ways to write n as a sum of 2, 3 or 4 squares (including zeros). The docstrings need the correct imports if this is added to SymPy:

```python
http://web.maths.unsw.edu.au/~mikeh/webpapers/paper73.pdf
from sympy.ntheory.primetest import is_square
def delta(n, den=None):
    if den:
        n, r = divmod(n, den)
        if r:
            return 0
    return is_square(n)

def d(r, m, n):
    return sum(1 for di in divisors(n) if di % m == r)

def e(r, s, m, n, den=None):
    if den and n % den:
        return 0
    return d(r, m, n) - d(s, m, n)

def p2(n):
    """Count the ways to write n as a sum of 2 or fewer squares.

    Examples
    ========

    >>> [p2(i) for i in range(1,18)]
    [1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1]
    >>> [len(list(sos(i,2,1))) for i in range(1,18)]
    [1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1]
    """
    return (e(1, 3, 4, n) + delta(n) + delta(n, 2))//2

def p3(n):
    """Count the ways to write n as 3 or fewer squares.

    Examples
    ========

    >>> [p3(i) for i in range(1,18)]
    [1, 1, 1, 1, 1, 1, 0, 1, 2, 1, 1, 1, 1, 1, 0, 1, 2]
    >>> [len(list(sos(i,3,1))) for i in range(1,18)]
    [1, 1, 1, 1, 1, 1, 0, 1, 2, 1, 1, 1, 1, 1, 0, 1, 2]
    """
    return (
        4*e(1, 3, 4, n) + 
        2*sum(e(1,3,4,n-k**2) for k in range(1, integer_nthroot(n, 2)[0] + 1)) + 
        3*e(1, 7, 8, n) + 
        3*e(3, 5, 8, n) + 
        5*delta(n) + 
        3*delta(n, 2) + 
        4*delta(n, 3))//12

def sigma(n):
    return sum(divisors(n, generator=True))

def sigma4(n):
    return sum(d for d in divisors(n, generator=True) if d % 4)

def sigma_not4(n):
    '''for i in range(1,10000):
        assert sigma_not4(i)==sigma4(i)
        assert sigma_not4(i)==sigma(i) - (4*sigma(i/4) if not i%4 else 0)
    '''
    if not n % 4:
        m = multiplicity(4, n)
        if m:
            n //= 4**m
            if n % 2:
                n *= 2
    return sum(divisors(n, generator=True))

def p4(n):
    """Count the ways to write n as 4 or less squares.

    Examples
    ========

    >>> [p4(i) for i in range(1,18)]
    [1, 1, 1, 2, 1, 1, 1, 1, 2, 2, 1, 2, 2, 1, 1, 2, 2]
    >>> [len(list(sos(i, 4, zeros=True))) for i in range(1,18)]
    [1, 1, 1, 2, 1, 1, 1, 1, 2, 2, 1, 2, 2, 1, 1, 2, 2]
    """
    N = integer_nthroot(n, 2)[0]
    N2 = integer_nthroot(n//2, 2)[0]
    return (
        sigma_not4(n) +
        17*e(1, 3, 4, n) + 
        4*sum(e(1,3,4,n - k**2) for k in range(1, N + 1)) +
        12*sum(e(1,3,4,n - 2*k**2) for k in range(1, N2 + 1)) +
        6*e(1, 7, 8, n) + 
        6*e(3, 5, 8, n) + 
        6*e(1, 3, 4, n, 2) + 
        8*e(1, 2, 3, n) + 
        16*e(1, 2, 3, n, 4) + 
        16*delta(n) + 
        12*delta(n, 2) + 
        8*delta(n, 3) + 
        12*delta(n, 4))//48
```

**Comment by smichr** [comment]
ping @thilinarmtb  -- ok, the new partitions is in place.

**Comment by thilinarmtb** [comment]
Great ! +1 to merge as it is.

File: sympy/utilities/iterables.py (hunk 1 above)
Comment (smichr) [inline sympy/utilities/iterables.py]: @jksuom, I just learned that the convention is to return count of 1 if a partition cannot be done -- one empty set -- and SymPy follows that convention:

```
>>> from sympy.functions.combinatorial.numbers import nT
>>> nT(0, 0)
1
```

So perhaps it's best to return an empty dictionary as you said so `sum(1 for 1 in partitions(n,k))` will return the right value regardless of `n` and `k`.

**Comment by smichr** [comment]
@thilinarmtb , did you write the `power_representation` routine? What's the backstory -- if you wrote it, was it just the most obvious way to you to solve the problem?
